### PR TITLE
feat: 1:1 donation matching

### PR DESCRIPTION
**Pull request by coszio**
# Context

Matching needs to be done so that the donations have a bigger effect and to motivate donors into giving money to the grants.

# Implementation

All the money from the funders will have to be transferred directly to the `program_info` account, which is a singleton for the program. From there, every time a `Donation` is created or incremented, it will be matched by money from the `program_info` account.

On every `Grant` creation, a "matching" `Donation` will be created (whose seeds will be different from the other `Donation`s) in which the total amount that has been matched to each grant is recorded. This is needed in case of a grant cancellation so that the correct amount of money goes back to the matcher as well.

### Edge Cases
When a donation is made (or incremented), the program will check that the grant is active, eligible for matching, and that the matcher has enough funds. Each case has its own error message.

### CODE REVIEW

File: programs/grants-program/src/state/donation.rs [hunk 1]
```diff
@@ -39,4 +39,11 @@ impl Donation {
     pub fn bump(&self) -> u8 {
         self.bump
     }
+
+    /// the equivalent of what will be matched with the current donation
+    pub fn matching_amount(&self) -> u64 {
+        // TODO: this will be replaced by quadratic matching or any other matching algorithm
+        // 1:1 matching
+        self.amount
+    }
```
Comment (coszio) [inline programs/grants-program/src/state/donation.rs:48]: FYI, this is where the algorithm for quadriatic matching can be implemented in the future

**Comment by coszio** [comment]
@jacobcreech Can you give this a look?

**Review by jpcaulfi — COMMENTED** [review]
Just FYI, you only need to include a `/// CHECK:` for `AccountInfo` - where the account is not being checked. Not needed on `Account<'info, T>`. (Noticed in `cancel_donation.rs`)   
   
However, I think the comments themselves are helpful. Just thought I'd mention it.   
   
Program-side looks good to me 👍    
   
I did notice an error when attempting to create a test grant on the UI - specifically in the "Sign Message" step. Not sure what the status is of the frontend. 

```shell
TypeError: Cannot read properties of undefined (reading 'fetch')
    at _callee$ (createGrant.ts?c607:19:69)
    at tryCatch (runtime.js?ecd4:45:16)
    at Generator.invoke [as _invoke] (runtime.js?ecd4:274:1)
    at prototype.<computed> [as next] (runtime.js?ecd4:97:1)
    at asyncGeneratorStep (createGrant.ts?c607:6:48)
    at _next (createGrant.ts?c607:6:48)
```

**Comment by coszio** [comment]
@jpcaulfi Thanks for checking it, we can disregard the UI for the grant creation for now. The whole flow is being improved by @yashikajotwani12 and @cdhiraj40 in #32

**Pull request merged**